pkg/execution/history/lifecycle.go

```diff
@@ -486,6 +487,126 @@ func (l lifecycle) OnWaitForEventResumed(
 	}
 }
 
+// OnInvokeFunction is called when a function is invoked from a step.
+func (l lifecycle) OnInvokeFunction(
+	ctx context.Context,
+	id state.Identifier,
+	item queue.Item,
+	op state.GeneratorOpcode,
+	eventID ulid.ULID,
+	corrID string,
+) {
+	logger.From(ctx).Debug().Interface("id", id).Msg("OnInvokeFunction")
+
+	groupID, err := toUUID(item.GroupID)
+	if err != nil {
+		l.log.Error(
+			"error parsing group ID",
+			"error", err,
+			"group_id", item.GroupID,
+			"run_id", id.RunID.String(),
+		)
+	}
+
+	fnID := ""
+	expiry := time.Time{}
+
+	opts, err := op.InvokeFunctionOpts()
+	if err != nil {
+		l.log.Error("error parsing invoke function options", "error", err)
+	}
+
+	if opts != nil {
+		fnID = opts.FunctionID
+		optsExp, err := opts.Expires()
+		if err != nil {
+			l.log.Error("error parsing invoke function options expiry", "error", err)
+		} else {
+			expiry = optsExp
+		}
+	} else {
+		l.log.Error("invoke function options are nil")
+	}
+
+	h := History{
+		ID:              ulid.MustNew(ulid.Now(), rand.Reader),
+		AccountID:       id.AccountID,
+		WorkspaceID:     id.WorkspaceID,
+		CreatedAt:       time.Now(),
+		FunctionID:      id.WorkflowID,
+		FunctionVersion: int64(id.WorkflowVersion),
+		GroupID:         groupID,
+		RunID:           id.RunID,
+		Type:            enums.HistoryTypeStepInvokingFunction.String(),
+		Attempt:         int64(item.Attempt),
+		IdempotencyKey:  id.IdempotencyKey(),
+		EventID:         id.EventID,
+		BatchID:         id.BatchID,
+		StepName:        &op.Name,
+		StepID:          &op.ID,
+		InvokeFunction: &InvokeFunction{
+			EventID:       eventID,
+			FunctionID:    fnID,
```

Comment:
`fnId` and `expiry` could have their zero values here. Should we set `InvokeFunction` to `nil` in that case? Or should `InvokeFunction.FunctionID` and `InvokeFunction.Timeout` be nullable?

Reply:
@goodoldneon Mm, I guess if `fnID` and `expiry` are zero values here then we should just return early instead of logging an error and continuing; they're required to be defined if we're calling this lifecycle hook.

Reply:
Maybe we should still write but set `InvokeFunction: nil` if we don't have all the data for `InvokeFunction`. Seeing the field as `null` in BigQuery would help troubleshooting

Reply:
@goodoldneon I'll add a check that we have all of the data before writing the log.

For the future, is it better for history in these cases to provide only everything/nothing, or is partial data also helpful? It'd lower the quality of the data overall, requiring readers to do more work, though in terms of helping debugging it might provide more threads to pull.